src/coreclr/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -2125,38 +2127,119 @@ void CodeGen::genFMAIntrinsic(GenTreeHWIntrinsic* node)
     regNumber op1Reg;
     regNumber op2Reg;
 
+    regNumber op1RegOld = op1->GetRegNum();
+    regNumber op2RegOld = op2->GetRegNum();
+    regNumber op3RegOld = op3->GetRegNum();
+
+    // For special case that targetReg equals to more than one op Regs
+    bool sameReg = false;
+    if ((op1RegOld == op2RegOld && targetReg == op1RegOld) || (op1RegOld == op3RegOld && targetReg == op1RegOld) ||
+        (op2RegOld == op3RegOld && targetReg == op1RegOld))
+        sameReg = true;
+
     bool       isCommutative   = false;
     const bool copiesUpperBits = HWIntrinsicInfo::CopiesUpperBits(intrinsicId);
 
     // Intrinsics with CopyUpperBits semantics cannot have op1 be contained
     assert(!copiesUpperBits || !op1->isContained());
 
-    if (op2->isContained() || op2->isUsedFromSpillTemp())
+    // Intrinsics with CopyUpperBits semantics must have op1 as target
+    if ((targetReg == op1RegOld || copiesUpperBits) && !sameReg)
```

Comment:
Can't we wrap this like this:

```c++
if (uniqueOverwrittenReg)
{
  if ((targetReg == op1RegOld)
  {
  }
  else if ((targetReg == op2RegOld)
  { 
  }
  else if ((targetReg == op3RegOld)
  { 
  }
  else
  {
     uniqueOverwrittenReg= false;
  } 
}

if (!uniqueOverwrittenReg)
{
    // For special case that targetReg is not same as any of the op*Reg
    // or targetReg is same for more than one op Regs
}
```

Reply:
@kunalspathak, yes, we could wrap the code like this. But this will add one more level of ```if``` condition. And also kind of "break" the code logic between the ```if (!uniqueOverwrittenReg)``` branch and the first three branches. I would prefer to keep current code logic unless you think it is better to be modified.

Reply:
> But this will add one more level of if condition.

True, my take is at least all the code related to the logic of "if there are unique operand with overwritten value" will be contained in one block. It will also help in changing the condition at one place rather than having to change `!sameReg` in 3 places and will be better code maintenance and ease readability.

It is a nit, so I will leave it to you to decide.

Let me know once you incorporate other feedback and we should go ahead once CI is green.

Reply:
@kunalspathak, I've resolved all the other comments in the latest code. For this specific change, I think your suggestion makes sense that all the code related to the logic will be in one block. I just kept the original code logic since both ways work fine from my perspective. 

Please take a look. Thank you very much!